src/transfer.jl

```diff
@@ -1,120 +1,210 @@
         
-        #calculate the index of the lowest layer the ray passes through
-        i = argmin(abs.(radii .- b))
+        #doing this with `findfirst` is messier at first and last index
+        i = argmin(abs.(radii .- b)) 
         if radii[i] < b
             i -= 1
         end
-        if i == 1 || i == 0
-            I[:, μ_ind] .= 0
-            continue
-        end #can eliminate for i == 1?
-        
-        #ds is the path length through each layer
-        ds = -diff(@. sqrt([R ; radii[1:i]]^2 - b^2)) 
-        dτ = α[1:i, :] .* ds
-        τ = cumsum(dτ, dims=1)
-        
-        for j in 1:size(τ, 2) #iterate over wavelength
-            I[j, μ_ind] = transfer_integral(τ[:, j], S[1:i, j]; plane_parallel=false)
+        lowest_layer_indices[μ_ind] = i
+
+        integrand_factor[μ_ind, 1:i] = @. radii[1:i] ./ sqrt(radii[1:i]^2 - b^2) * τ_ref_α_ref[1:i]
+    end
+
+    #preallocations
+    I = Matrix{el_type}(undef, size(α, 2), length(μ_surface_grid)) #the surface intensity
+    integrand = Vector{el_type}(undef, size(α, 1))                 #integrand of τ integral 
+    τ_λ = Vector{el_type}(undef, size(α, 1))                       #optical depth at a particular λ
+    for λ_ind in 1:size(α, 2), μ_ind in 1:length(μ_surface_grid)
+        i = lowest_layer_indices[μ_ind]
+        for k in 1:i #I can't figure out how to write this as a fast one-liner
+            integrand[k] = α[k, λ_ind] * integrand_factor[μ_ind, k]
         end
-        if b > r0
-            #if the ray never leaves the model atmosphere, include 
-            #the contribution from the other side of the star
-            τ_prime = τ[end:end, :] .+ [cumsum(reverse(dτ[2:end, :], dims=1), dims=1) 
-                                        ; zeros(size(τ[end:end, :]))]
-            for j in 1:size(τ, 2) #Iterate over wavelength
-                I[j, μ_ind] += transfer_integral(τ_prime[:, j], S[1:i, j]; plane_parallel=false)
-            end
-        else #otherwise assume I=S at atmosphere lower boundary
-            I[:, μ_ind] += @. exp(-τ[end, :]) * S[end, :]
+        cumulative_trapezoid_rule!(τ_λ, log_τ_ref, integrand, i)
+        I[λ_ind, μ_ind] = ray_transfer_integral(view(τ_λ,1:i), view(S,1:i,λ_ind))
+
+        if i < length(radii)
+            #if the ray never leaves the model atmosphere, include the contribution from the 
+            #other side of the star.  Calculate the optical depths you get by reversing the τ_λ.
+            #This is less accurate than actually integrating to find τ, but the effect is small.
+            #This should probably by audited for off-by-one errors.  It's also inneficient.
+            τ_prime = τ_λ[i] .+ [cumsum(reverse(diff(view(τ_λ,1:i)))) ; 0]
+            I[λ_ind, μ_ind] += ray_transfer_integral(view(τ_prime, 1:i), view(S,1:i,λ_ind))
+        else #otherwise assume I=S at atmosphere lower boundary.  This is a _tiny_ effect.
+            I[λ_ind, μ_ind] += exp(-τ_λ[end]) * S[end, λ_ind]
         end
```

Comment:
It doesn't seem like branch-misses are a problem (then again, it's never been clear to me how they would manifest in the flamegraph), but I'm pretty sure that this branch does not need to exist within the innermost loop.

If it ever became necessary, I think you could subdivide `for μ_ind in 1:length(μ_surface_grid)` into 2 separate branches that execute the different branches.

Reply:
This is an excellent point.

Reply:
I've noted it in a comment.